Expand Open a Realm page (#1272)
mongodb/docs-realm: source/sdk/dotnet/examples/open-a-realm.txt

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The first time you open a synced {+realm+}, the app must be able to connect to 
your {+service-short+} backend. Your app code walks the user through 
:ref:`authenticating <dotnet-authenticate>` first. Once you have a 
:dotnet-sdk:`User <reference/Realms.Sync.User.html>` object, you open a synced 
{+realm+} by calling the 
:dotnet-sdk:`GetInstanceAsync() <reference/Realms.Realm.html#Realms_Realm_GetInstanceAsync_Realms_RealmConfigurationBase_System_Threading_CancellationToken_>` 
method, passing in a 
:dotnet-sdk:`SyncConfiguration <reference/Realms.Sync.SyncConfiguration.html>` 
object that includes the :ref:`partition name <partition-value>` and 
the :dotnet-sdk:`User <reference/Realms.Sync.User.html>` object. The following 
code demonstrates this:

.. literalinclude:: /examples/generated/code/start/Examples.codeblock.open-synced-realm.cs
   :language: csharp

In the above example, the code shows how to open the {+realm+} *asynchronously* 
by calling ``GetInstanceAsync()``. You can also open a {+realm+} *synchronously* 
by calling the 
:dotnet-sdk:`GetInstance() <reference/Realms.Realm.html#Realms_Realm_GetInstance_System_String_>` 

> method:

Reviewer: I know that iOS never returns a realm for an async open when offline. Android, on the other hand, is happy to return an empty realm, even if you've never opened it before (unless you specify a waitForInitialData option in the sync configuration, and even that has a timeout). Does .NET have any limitations like this? Could be worth mentioning here or in the next section.

.. literalinclude:: /examples/generated/code/start/Examples.codeblock.open-synced-realm-sync.cs
   :language: csharp

.. seealso::

   :ref:`Partitions <sync-partitions>`

   :ref:`Partition Strategies <partition-strategies>`

Open a Synced Realm While Offline
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Once a user authenticates on a device, the User object is persisted until the 

Reviewer: I wonder if we should put User in monospace here since we're talking about the programmatic object. Also, passive voice/moving "on that device" to demonstrate the persistence occurring on the device as opposed to in da cloud:
```suggestion
Once a user authenticates, the ``User`` object persists on that device until the 
```

> User logs off. This allows your app to 

Reviewer: ```suggestion
user logs off. This allows your app to 
```
And here I'd say we're referring to the physical (probably human) user of the app, who we can just use the pronoun for. I think.

:ref:`retrieve an existing user <dotnet-retrieve-current-user>` and open a 

> synced realm in an offline state. Changes that occur while offline will be 

Reviewer: I'll never stop nitpicking the source constant omissions :) 

```suggestion
synced {+realm+} in an offline state. Changes that occur while offline will be 
```

synced by {+service-short+} once the device is reconnected.

The following code shows how to check if there is an existing ``User`` object. 
If none is found, it uses the process outlined about to obtain a user. If the 
device already have a ``user``, it opens the synced {+realm+} with that user:

.. literalinclude:: /examples/generated/dotnet/Examples.codeblock.check-if-offline.cs
   :language: csharp

.. note::

>    The first time a user logs on to your realm app, you should open the realm 

Reviewer: ```suggestion
   The first time a user logs on to your app, you should open the {+realm+}
```
I think we can just use "app" here since the user doesn't have to be aware of realm and we normally reserve the "realm app" term for the backend, even though that's also technically correct here. And a source constant.

>    *asynchronously* to sync data from the server to the device. After that initial 

Reviewer: I think we should pull this information directly in to the opening paragraph of "open a synced realm while offline", and possibly justify it more. Will opening a realm asynchronously while offline cause the realm to never open or crash the app? What happens if I open the realm synchronously when I first open it -- just a chuggy, slow app while it downloads changes, or does it fail or cause issues? I think this note is easy to ignore and the opening paragraph of this section would provide a better place to explain these tradeoffs.

.. literalinclude:: /examples/generated/code/start/Examples.codeblock.open-synced-realm.cs
   :language: csharp

In the above example, the code shows how to open the {+realm+} *asynchronously* 
by calling ``GetInstanceAsync()``. You can also open a {+realm+} *synchronously* 
by calling the 
:dotnet-sdk:`GetInstance() <reference/Realms.Realm.html#Realms_Realm_GetInstance_System_String_>` 
method:

.. literalinclude:: /examples/generated/code/start/Examples.codeblock.open-synced-realm-sync.cs
   :language: csharp

.. seealso::

   :ref:`Partitions <sync-partitions>`

   :ref:`Partition Strategies <partition-strategies>`

Open a Synced Realm While Offline
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Once a user authenticates, the ``User`` object persists on the device until the 
user logs off. This allows your app to 
:ref:`retrieve an existing user <dotnet-retrieve-current-user>` and open a 
synced {+realm+} in an offline state. Changes that occur while offline will be 
synced by {+service-short+} once the device is reconnected.

The following code shows how to check if there is an existing ``User`` object. 
If none is found, it uses the process outlined about to obtain a user. If the 

> device already have a ``user``, it opens the synced {+realm+} with that user:

Reviewer 2: ```suggestion
device already has a ``user``, it opens the synced {+realm+} with that user:
```
